### PR TITLE
Add patch to skip using collision poly check table (performance optimization)

### PR DESCRIPTION
**Pull request by Hexocyte**
I ported a decomp-based patch HackerN64/HackerOoT#1 to an ASM hack.

This table is used to skip checking polygons that have already been checked in the same line test, but it's slower than just rechecking them, particularly in large scenes.

### Testing
I timed running across the Fire Temple lobby in three emulators. I recorded video and did the timing based on the video and checked how low the game FPS went for the emus that show that.

<table>
<thead>
<tr>
<td></td>
<th colspan="5">Unpatched — time in seconds (/ low fps)</th>
<th colspan="5">Patched — time in seconds (/ low fps)</th>
</tr>
<tr>
<td></td>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
</tr>
</thead>
<tbody>
<tr>
<td>Ares v131</td>
<td>6.88</td>
<td>6.88</td>
<td>6.78</td>
<td>6.87</td>
<td>7.05</td>
<td>6.15</td>
<td>6.25</td>
<td>6.25</td>
<td>6.12</td>
<td>6.12</td>
</tr>
<tr>
<td>Project64 3.0.1.5664 (interpreter) in Wine</td>
<td>6.08 / 16.0</td>
<td>5.97 / 16.4</td>
<td>5.93 / 16.7</td>
<td>5.97 / 16.3</td>
<td>5.98 / 16.2</td>
<td>5.62 / 19.3</td>
<td>5.63 / 19.0</td>
<td>5.62 / 19.2</td>
<td>5.63 / 18.9</td>
<td>5.62 / 19.5</td>
</tr>
<tr>
<td>Dolphin 5.0-17995</td>
<td>6.75 / 14.01</td>
<td>6.70 / 13.91</td>
<td>6.68 / 14.05</td>
<td>6.63 / 13.82</td>
<td>6.65 / 13.82</td>
<td>6.63 / 14.11</td>
<td>6.65 / 14.09</td>
<td>6.63 / 14.22</td>
<td>6.58 / 14.30</td>
<td>6.67 / 14.03</td>
</tr>
</tbody>
</table>

These results show significant speedup from the patch.

To test how this performs in small scenes, I used gz. I spawned 40 spike traps in Morpha's room, and edited their update code using gz's memory editor to increase the length of their line check from 30 to 300. I then added this patch using the memory editor and compared how fast it ran before and after the patch. In PJ64 it went from 11.0 DL/s to 11.3 DL/s. Ares went from 9.42 fps to 9.55 fps (timed using a video). From this it seems the patch is still a tiny (perhaps negligible) optimization in small scenes.

### CODE REVIEW

**Comment by ETR-BTF** [comment]
Does the speed-up apply to console too?

**Comment by cjohnson57** [comment]
Seems good to me although I want a second pair of eyes on it before I merge

**Comment by Hexocyte** [comment]
> Does the speed-up apply to console too?

I expect it would on N64 since I'm guessing the HackerN64 devs checked that their patch improved N64 performance; albeit being decomp-based, theirs is bound to be better than mine. And this removed a lot of lag in my emulator Fire Temple timings, but I suppose it's possible the emulators are wrong.

I'm less sure about Wii because it didn't make as much difference on Dolphin, and I don't think Dolphin is as accurate timingwise.

My intuition is it should be faster to sometimes recheck a few polygons than to reset the table by looping for every polygon in a large scene. The polygons will get rechecked if they occupy multiple of the same subdivisions as the line being checked, but not every line check will stretch between subdivisions, and the largest scene is Ganon's Castle with 5096 polygons.

**Review by cjohnson57 — APPROVED** [review]

**Pull request merged**
